pkg/icingadb/history/sorter.go

```diff
@@ -0,0 +1,402 @@
+	verbose bool
+}
+
+// NewStreamSorter creates a StreamSorter honoring the given context and returning elements to the callback function.
+func NewStreamSorter(
+	ctx context.Context,
+	logger *logging.Logger,
+	callbackFn func(msg redis.XMessage, args any) bool,
+) *StreamSorter {
+	sorter := &StreamSorter{
+		ctx:               ctx,
+		logger:            logger,
+		callbackFn:        callbackFn,
+		submissionCh:      make(chan streamSorterSubmission),
+		closeChSubmission: make(chan chan<- redis.XMessage),
+		closeChQueue:      make(chan chan<- redis.XMessage),
+	}
+
+	_ = context.AfterFunc(ctx, func() {
+		close(sorter.submissionCh)
+		close(sorter.closeChSubmission)
+		close(sorter.closeChQueue)
+	})
+
+	ch := make(chan streamSorterSubmission)
+	go sorter.submissionWorker(ch)
+	go sorter.queueWorker(ch)
+
+	return sorter
+}
+
+// submissionWorker listens ton submissionCh populated by Submit, fills the heap and ejects streamSorterSubmissions into
+// out, linked to the queueWorker goroutine for further processing.
+func (sorter *StreamSorter) submissionWorker(out chan<- streamSorterSubmission) {
+	defer close(out)
+
+	// When a streamSorterSubmission is created in the Submit method, the current time.Time is added to the struct.
+	// Only if the submission was at least three seconds (submissionMinAge) ago, a popped submission from the heap will
+	// be forwarded to the other goroutine for future processing.
+	const submissionMinAge = 3 * time.Second
+	submissionHeap := &streamSorterSubmissions{}
+
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-sorter.ctx.Done():
+			return
+
+		case sub, ok := <-sorter.submissionCh:
+			if !ok {
+				return
+			}
+
+			if sorter.verbose {
+				sorter.logger.Debugw("Push submission to heap", zap.Object("submission", sub))
+			}
+
+			heap.Push(submissionHeap, sub)
+
+		case ch, ok := <-sorter.closeChSubmission:
+			if !ok {
+				return
+			}
+
+			bkp := &streamSorterSubmissions{}
+			for submissionHeap.Len() > 0 {
+				x := heap.Pop(submissionHeap)
+				sub, ok := x.(streamSorterSubmission)
+				if !ok {
+					panic(fmt.Sprintf("invalid type %T from submission heap", x))
+				}
+
+				if sub.out == ch {
+					continue
+				}
+
+				bkp.Push(sub)
+			}
+			submissionHeap = bkp
+
+		case <-ticker.C:
+			start := time.Now()
+			submissionCounter := 0
+
+			for submissionHeap.Len() > 0 {
+				x := heap.Pop(submissionHeap)
+				sub, ok := x.(streamSorterSubmission)
+				if !ok {
+					panic(fmt.Sprintf("invalid type %T from submission heap", x))
+				}
+
+				if time.Since(sub.submitTime) < submissionMinAge {
+					if sorter.verbose {
+						sorter.logger.Debugw("Stopped popping heap as submission is not old enough",
+							zap.Object("submission", sub),
+							zap.Int("submissions", submissionCounter),
+							zap.Duration("duration", time.Since(start)))
+					}
+
+					heap.Push(submissionHeap, sub)
```

Comment:
It shouldn't be necessary to pop the element to take a peek at it:

> The minimum element in the tree is the root, at index 0.

-- https://pkg.go.dev/container/heap